### DISCUSSION TITLE
A hole in a rectangle

### DISCUSSION THREAD

**Discussion by inveders** [google/filament#4210] [Q&A] [answered]
Hello,
I would like to make one or more holes in a rectangle drawn with Filament. Let's take the simplest example 1 hole. For this I tried the first possibility, to draw only in the depth buffer an invisible rectangle (which will represent the hole) and to draw in the color buffer, superimposed on the first one, a visible rectangle, bigger.

**Here is the .mat file of my invisible rectangle**

````
material {
    name : cutting,
    parameters : [
            {
                type : float4,
                name : baseColor
            },
            {
                type : float,
                name : roughness
            }
        ],
    requires : [
        color
    ],
    shadingModel : unlit,
    culling : none,
    depthWrite : false,
    colorWrite : false,
    depthCulling : false,
    shadowMultiplier : true,
    blending : transparent


}

fragment {
    void material(inout MaterialInputs material) {
        prepareMaterial(material);
        material.baseColor = materialParams.baseColor;
    }
}
````
By the way I think that the required color and parameters are useless here

**Here is the .mat file of my visible rectangle**

````
material {
    name : opaque_square,
    parameters : [
            {
                type : float4,
                name : baseColor
            },
            {
                type : float,
                name : roughness
            }
        ],
    requires : [
        color
    ],
    shadingModel : unlit,
    culling : none
}

fragment {
    void material(inout MaterialInputs material) {
        prepareMaterial(material);
        material.baseColor = materialParams.baseColor;
    }
}
`````
Logic tells me that without any other action on my part (multiplicity, superposition or other, reverse the order of the drawings), it is quite normal for the visible rectangle to "hide" the invisible one. How could I create this hole this way?

The second option is to use a CSG (Constructive Solid Geometry) library to subtract the two rectangles (I specify that here I am in 2D since the Z coordinates of my rectangles are identical). I heard about OpenCSG but I don't know how to integrate it in Android Studio and how to generate vertices from it to create my rectangle directly with the right coordinates taking into account the hole.

Do you have any advice on theses points?

**Comment by jwwalker** [comment]
You said that the invisible rectangle would draw only in the depth buffer, but its material says `depthWrite : false`, so it's not drawing in the depth buffer.  That said, this scheme would only work if you can guarantee that the invisible rectangle draws before the visible one, and I don't think Filament gives control over the order in which things are drawn.

Another possibility would be to draw just one big rectangle, but with a texture with an alpha channel, and punch out the holes in the texture.

**Reply by romainguy** [reply]
You can control ordering by setting priorities.

**Reply by inveders** [reply]
Thank you both for your feedback, @romainguy when you talk about setting priorities, where can I handle it? In the .mat file? The order of creation of my rectangles? Is a relative topic about it in the filament doc?

**Reply by romainguy** [reply]
It’s a feature of the renderables: https://github.com/google/filament/blob/f4558fee1d5707459e095c97cef179cdd6621bc7/android/filament-android/src/main/java/com/google/android/filament/RenderableManager.java#L252

You could do also use an alpha mask material like @jwwalker suggested.

**Reply by inveders** [reply]
thanks to you I was able to change the depthWrite to true and set the priorities at the renderable level (the invisible rectangle to 0 and the visible rectangle to 4 or 7), remove the shadowMultiplier and blending too. With this modifications I have a "black rectangle in the middle of my rectangle, great) but it's not a real transparent hole. This is my new .mat file. When I add the blending to transparent there is no more rectangle (it's normal it's transparent). So how to have a hole? According to you, where is the issue?

```
material {
    name : cutting,
    parameters : [
            {
                type : float4,
                name : baseColor
            },
            {
                type : float,
                name : roughness
            }
        ],
    requires : [
        color
    ],
    shadingModel : unlit,
    culling : none,
    depthWrite : true,
    colorWrite : false,
    depthCulling : false

}

fragment {
    void material(inout MaterialInputs material) {
        prepareMaterial(material);
        material.baseColor = materialParams.baseColor;
    }
}
```
For wich rectangle (invisible or opaque) the alpha mask material is ?

**Reply by romainguy** [reply]
So, forget about using depth, just focus on using alpha masking, it will be easier. For alpha masking, you just need to set the blend mode to "masked". By default, any pixel with an alpha value < 0.4 will become a hole. So in your material you can simply write `material.baseColor = texture(materialParams_myTexture, getUV0());`, and draw the hole in the texture's alpha channel. If you do not need color from the texture, you could do something simpler and write `material.baseColor.a = texture(materialParams_myTexture, getUV0()).r;` for instance, where the hole would come from the *red* channel of the texture.

Note: this is typically how foliage or other objects (fences, chains, etc.) with transparent cutouts work.

Another note: with the priorities you are probably seeing black because you are interfering with how your background is rendered (which also has a priority and respect the depth buffer by default). For your trick to always work you'd need to draw the background *first* (prob with depth testing off).

**Reply by jwwalker** [reply]
@romainguy When using masked blending, is it not necessary to unpremultiply the color for fragments that are not discarded, or does Filament handle that automatically?

Regarding your suggestion of drawing the background first, with depth testing off:   In the discussion of `depthCulling` in the materials guide, it says "When depth testing is disabled, an object rendered with this material will always appear on top of other opaque objects."  So, that's not actually true, right?  How does `depthCulling` interact with rendering priorities?

**Reply by romainguy** [reply]
You should not premultiply RGB when using a masked material.

depthCulling = depth testing. So if you draw a full screen rectangle last with depth testing off, everything drawn before will be covered. Hence why you’d need to draw it first.

**Reply by jwwalker** [reply]
OK, so it would be correct to say "When depth testing is disabled, an object rendered with this material will always appear on top of other opaque objects of the same or lower priority."?

**Reply by inveders** [reply]
Ok @romainguy @jwwalker to make sure I got it right with alpha mask material, I made a [little sample based on the hello-triangle sample](https://we.tl/t-ZUdERfDoir).  I added a wooden texture to the first triangle and added a second smaller triangle on top (which is supposed to represent the hole). I used a masked material for the small triangle to make a hole in the first one. What I didn't understand is if both triangles have to have the same material? Or does the large triangle have to have a non-blending material? I'm a little confused about the order of things. The result of the sample is as follows: on a gray background, my large triangle with a wood texture, and the second triangle all white above, however I do not see a hole (what I would like to see is the gray background behind and not the white triangle)

**Reply by jwwalker** [reply]
No, with the alpha mask technique, you do not need two triangles, just the bigger one, which should have a masked material.  And the texture should be wood (or whatever you like) with a transparent "hole" cut out of it.

**Reply by inveders** [reply]
@jwwalker Hum I'm sorry I don't understand how to make the hole in the triangle, ok for the alpha masked texture of the big triangle, but for the rest it's really blurry. How to determine the size of the hole in my triangle without creating a second one?

**Reply by jwwalker** [reply]
You make a hole in the texture, not the geometry.  That is, make part of the image totally transparent.  If you don't know how to do that, check the documentation of whatever image editing program you use.  I don't know what you mean by "hidden texture of the big triangle", why would it be hidden?

The image below has a transparent triangle, though you can't really tell that it's transparent with the solid background here.

![cherry-cutout](https://user-images.githubusercontent.com/730426/124518688-ee24f180-dd9b-11eb-8160-a04e54ba699f.png)

**Reply by inveders** [reply]
@jwwalker  In fact I need to create a wall (my rectangle) and make holes in it dynamically (with a setOnTouchListener) and be able to see behind my wall. Let me explain, I am in my living room (I use ArCore and Filament) and I shoot the wall of the living room. I create a virtual wall with my rectangle, but I have photo frames on the wall, so I will create holes on my rectangle to be able to see these photo frames. That's what I'm trying to do, but I have the impression here that having to create the hole in the .mat file doesn't allow me to do what I want dynamically.

**Reply by jwwalker** [reply]
You wouldn't create a hole in the .mat file, but in the texture, which would be a parameter supplied to the `MaterialInstance`, and could be changed dynamically.  However I don't understand why you would want holes for photo frames, rather than having the photo frames being objects in front of the wall.

**Reply by inveders** [reply]
@jwwalker It's to permit to the user to choose if if want to see them or not. Please do you have a code example of your hole ?
Because if I can create a hole with a setParameter in my materialInstance, so I have to know which type of parameter it is and how to set the dimensions or size of my hole.

I made a mistake with the "hidden texture", it's the alpha mask texture sorry.

**Reply by jwwalker** [reply]
I did not say that a hole is a parameter, I said that the texture is a parameter, and the texture image might have a hole.  The example I showed of an image with a hole was made in an image editing program.  While one could put holes in an image programmatically, I have never done that.

If a picture frame was an object in front of the wall, then you could show them or not by adding renderables to the scene or removing them from the scene.

**Reply by inveders** [reply]
Oh OK @jwwalker  now I understand. In my case my texture parameter set a basecolor with a drawable (my wood png) without hole. And according to your comment I would have a hole beeing part of my image. So it's definitively not the good method for me because I draw my wall by geometry, dynamically in a 3D space, and I need to set the hole in specific coordinates, and I think that it's not possible with an other software image editing...

I take picture frame like example but it's could be all kind of things which could be on my wall (tv, etc) and theses things are real object from my camera and not renderable objects I create, so it's difficult to remove or add them since create more renderables.

Finally I don't know if filament permit to do what I want in this specific case 😔 Maybe with the CSG or the first method with the two rectangles, one of top of the other (I think that I will return to the first method)
@romainguy do you understand my issue? Have you a method to propose about this case?

**Comment by inveders** [comment, marked as answer]
Hey everyone, I finally figured out how to do the first method (layering two different sized rectangles to create a hole in the larger rectangle).  Thank you for all your answers which helped me a lot. So here is the solution: 
- Firstly, don't forget to specify that the camera must have priority over the two rectangles (priority 0 or 1 for example)
- Draw the invisible rectangle, the smaller one, in second thanks to the priorities (priority 2 for example), in depthBuffer with depthWrite to true, and set the depthCulling to true.
- Draw the rectangle, the largest, in which you want to make the hole, in third position thanks to the priorities (priority 7 for example)

The two 2D rectangles must of course be superimposed in the 3D environment if this is your case (pay attention to the coordinates especially on the Z axis)

Here it is, with all these points, I have my hole displayed, I was only missing the priority on the camera and the depthCulling last time.

To finish, here are the two .mat files of my rectangles

The invisible rectangle
```kotlin
material {
    name : invisible,
    shadingModel : unlit,
    colorWrite : false,
    depthWrite : true,
    depthCulling : true,
    culling : none,

       parameters : [
           {
               type : sampler2d,
               name : myTexture
           }
       ],
       requires : [
           uv0
       ]
   }
   fragment {
       void material(inout MaterialInputs material) {
           prepareMaterial(material);
           material.baseColor.a = texture(materialParams_myTexture, getUV0()).r;
       }
   }
````

The visible rectangle
```kotlin
material {
    name : finalwall,
    shadingModel : unlit,
    culling : none,
    parameters : [
        // Base color sRGB texture
        {
            type : sampler2d,
            name : myTexture
        }
    ],
    // To sample textures our material must declare that it requires
    // a set of UV coordinates from the rendered mesh
    requires: [
        uv0
    ]
}

fragment {
    void material(inout MaterialInputs material) {
        prepareMaterial(material);

        material.baseColor = texture(materialParams_myTexture, getUV0());

    }
}
````
Good luck with this
